pkg/logs/internal/launchers/file/provider/file_provider.go

```diff
@@ -0,0 +1,281 @@
+			wildcardFileCounter.setTotal(source, len(files))
+			if err != nil {
+				continue
+			}
+			wildcardFiles = append(wildcardFiles, files...)
+		}
+
+		p.applyOrdering(wildcardFiles)
+		consumeFiles(wildcardFiles)
+	}
+
+	// Record what ratio of files each wildcard source tracked
+	for source, matchCnt := range wildcardFileCounter.counts {
+		source.Messages.AddMessage(source.Config.Path, fmt.Sprintf("%d files tailed out of %d files matching", matchCnt.tracked, matchCnt.total))
+	}
+
+	if len(filesToTail) == p.filesLimit {
+		log.Warn("Reached the limit on the maximum number of files in use: ", p.filesLimit)
+		return filesToTail
+	}
+
+	return filesToTail
+}
+
+// CollectFiles takes a 'LogSource' and produces a list of tailers matching this source
+// with ordering defined by 'sortMode'
+func (p *FileProvider) CollectFiles(source *sources.LogSource) ([]*tailer.File, error) {
+	path := source.Config.Path
+	_, err := os.Stat(path)
+	switch {
+	case err == nil:
+		return []*tailer.File{
+			tailer.NewFile(path, source, false),
+		}, nil
+	case config.ContainsWildcard(path):
+		files, err := p.filesMatchingSource(source)
+		if err != nil {
+			return nil, err
+		}
+		p.applyOrdering(files)
+
+		return files, err
+	default:
+		return nil, fmt.Errorf("cannot read file %s: %s", path, err)
+	}
+}
+
+// filesMatchingSource returns all the files matching the source path pattern.
+func (p *FileProvider) filesMatchingSource(source *sources.LogSource) ([]*tailer.File, error) {
+	pattern := source.Config.Path
+	paths, err := filepath.Glob(pattern)
+	if err != nil {
+		return nil, fmt.Errorf("malformed pattern, could not find any file: %s", pattern)
+	}
+	if len(paths) == 0 {
+		// no file was found, its parent directories might have wrong permissions or it just does not exist
+		return nil, fmt.Errorf("could not find any file matching pattern %s, check that all its subdirectories are executable", pattern)
+	}
+
+	excludedPaths := make(map[string]int)
+	for _, excludePattern := range source.Config.ExcludePaths {
+		excludedGlob, err := filepath.Glob(excludePattern)
+		if err != nil {
+			return nil, fmt.Errorf("malformed exclusion pattern: %s, %s", excludePattern, err)
+		}
+		for _, excludedPath := range excludedGlob {
+			log.Debugf("Adding excluded path: %s", excludedPath)
+			excludedPaths[excludedPath]++
+			if excludedPaths[excludedPath] > 1 {
+				log.Debugf("Overlapping excluded path: %s", excludedPath)
+			}
+		}
+	}
+
+	files := make([]*tailer.File, 0, len(paths))
+	for _, path := range paths {
+		if excludedPaths[path] == 0 {
+			files = append(files, tailer.NewFile(path, source, true))
+		}
+	}
+
+	return files, nil
+}
+
+// applyOrdering sorts the 'files' slice in-place by the currently configured 'sortMode'
+func (p *FileProvider) applyOrdering(files []*tailer.File) {
+	if p.wildcardOrder == WildcardMtime {
+		// sort paths descending by mtime
+		sort.SliceStable(files, func(i, j int) bool {
+			statI, _ := os.Stat(files[i].Path)
```

Comment:
Would it be more efficient (and possibly more performant) to first call stat on all the files, and then sort them based on the results? 

This will call `os.Stat` at least twice for every file as is.

Reply:
Good point! Will stat them all first and then sort over the results